### ISSUE TITLE
[api-extractor] .d.ts rollup sometimes emits names that conflict with global names

### ISSUE THREAD

**Issue by octogonz** [microsoft/rushstack#1095]
Consider this example:

```ts
import { Promise as MyPromise } from './file1';

export function f(p1: Promise<void>, p2: MyPromise<void>): void {
  // p1 is using the ambient Promise from the compiler runtime
  // p2 is using the declaration from file1, which happens to use the same name "Promise"
  // in that file
}
```

In the .d.ts rollup, `p1` and `p2` accidentally end up with the same type:
```ts
declare class Promise<T> {
    private x;
}

export declare function f(p1: Promise<void>, p2: Promise<void>): void;
```

This happens because the algorithm doesn't realize that `Promise` is a global name introduced by **lib.es2015.promise.d.ts**, which comes in via this part of **tsconfig.json**:

```ts
    "lib": [
      "es5",
      "scripthost",
      "es2015.collection",
      "es2015.promise",
      "es2015.iterable",
      "dom"
    ],
```

This is breaking **@microsoft/sp-application-base**.

@iclanton

**Comment by octogonz** [comment]
In **typescript.d.ts** I found this interface, which checks whether a name is global or not:

```ts
    interface PrintHandlers {
        /**
         * A hook used by the Printer when generating unique names to avoid collisions with
         * globally defined names that exist outside of the current source file.
         */
        hasGlobalName?(name: string): boolean;
```

However, it is a contract meant to be implemented by the caller of the `Printer`.  Internally the compiler seems to implement it via [the EmitResolver](https://github.com/Microsoft/TypeScript/blob/45899ad3d9d15018cee37c8e8507342cd8b44ca3/src/compiler/emitter.ts#L188), which has [this code](https://github.com/Microsoft/TypeScript/blob/master/src/compiler/checker.ts):

```ts
namespace ts {
    . . .
    export function createTypeChecker(host: TypeCheckerHost, produceDiagnostics: boolean): TypeChecker {
        . . .
        const globals = createSymbolTable();
        . . .
        const checker: TypeChecker = {
            . . .
            hasGlobalName
            . . .
        };
        . . .
        return checker;
        . . .
        function hasGlobalName(name: string): boolean {
            return globals.has(escapeLeadingUnderscores(name));
        }
```

But when I try to use `program.getTypeChecker()` it complains that it was created with `produceDiagnostics = false`.

**Comment by octogonz** [comment]
The `"es2015.promise"` module can be found in the list of processed source files via `this.program.getSourceFiles()`, but these modules don't have symbols, so there's no "exports" that we can easily retrieve from them.

From any source file, you can enumerate `Promise` and `Map` and most other global names like this:

```ts
this.typeChecker.getSymbolsInScope(sourceFile, ts.SymbolFlags.Transient)
```
...however I'm pretty sure there exist global symbols that are not "transient", as well as  "transient" symbols that are not global names.

The compiler docs don't say much about what "transient" actually means besides this, but it sounds like the wrong concept for what we need:
```ts
    export const enum SymbolFlags {
        . . .
        Transient               = 1 << 25,  // Transient symbol (created during type check)
```

@RyanCavanaugh @DanielRosenwasser any idea how to find the global names?  For example, is there any example anywhere of someone implementing the `PrintHandlers.hasGlobalName` contract mentioned above?

**Comment by octogonz** [comment]
Okay, I found several internal APIs that can be combined together to reach `hasGlobalName()`:  :-)

```ts
const typeChecker = (this.program as any).getDiagnosticsProducingTypeChecker();
const resolver = (typeChecker as any).getEmitResolver();
resolver.hasGlobalName('Promise'); // --> true
```

Is this a valid/safe solution?

**Comment by octogonz** [comment]
Hmm... the symbol table being queried by `hasGlobalName()` has 1650 names in it... is that really correct?

Besides `Promise`, it also finds:
- `Partial`
- `RTCStatsIceCandidateType`
- `SVGAnimatedInteger`
- `encodeURI`
- `eval`
- `it`
- `"https"` (quotes included)
- `"v8"`

**Comment by RyanCavanaugh** [comment]
@octogonz that's correct. The global symbol table has everything you could reach from an unqualified position (Partial, encodeURI, etc) as well as all module names you could import without path qualification ("https", etc).

**Comment by octogonz** [comment]
@RyanCavanaugh Welcome back, BTW!  😄

**Comment by octogonz** [comment]
> ```ts
> const typeChecker = (this.program as any).getDiagnosticsProducingTypeChecker();
> const resolver = (typeChecker as any).getEmitResolver();
> resolver.hasGlobalName('Promise'); // --> true
> ```

@rbuckton @RyanCavanaugh 

I'm looking at implementing this now.  Is `getDiagnosticsProducingTypeChecker()` still the most reasonable way to access `hasGlobalName()`?

I noticed this slightly concerning comment:
```ts
        // For testing purposes only.  Should not be used by any other consumers (including the
        // language service).
        /* @internal */ getDiagnosticsProducingTypeChecker(): TypeChecker;
```

**Comment by octogonz** [comment]
Fixed by PR #1767 which was published with API Extractor 7.7.12.
